### PR TITLE
Add ScaleLayer and standardize function

### PR DESCRIPTION
**Pull request by craffel**
The motivation behind this layer is this: Usually, part of your ML pipeline is to compute some statistics on the training set (mean and standard deviation, or min and max) and then use them to normalize/standardize all inputs to the network.  Currently, this needs to be done "outside of the model", i.e. when you save your model you must also save the training set statistics to use later when passing data through the model.  There's no convenient way to store this information in the model itself.  This layer simply stores these statistics, and then applies them in `get_output` in a smart way which hopefully covers most use-cases.  See the example in the docstring for an example.

As another (personally relevant) example, say you are passing sequences of feature tensors into a convolutional network.  Your input data is of shape `(n_batch, n_filters, n_time_steps, n_feature)`.  You want to normalize over the `n_filters` and `n_feature` (feature) dimensions, not the `n_batch` or `n_time_steps` (sample) dimensions.  After reshaping/dimshuffling your training data (call it `X`) to shape `(n_batch*n_time_steps, n_filters, n_feature)`, you could compute the mean and standard deviation over these dimensions as `offset = X.mean(axis=0)`, `scale = X.std(axis=0)` and then apply them using the `StandardizationLayer` as follows:

```Python
l_in = lasagne.layers.InputLayer((None, n_filters, None, n_features))
l_std = lasagne.layers.normalization.StandardizationLayer(
    l_in, scale, offset, axis=(1, 3))
```

If this makes sense and seems useful, I will create some unit tests for it and/or add/remove/change features.

### CODE REVIEW

**Comment by f0k** [comment]
+1 for adding a layer that just scales and shifts, this is currently not easily possible. The way it's presented might evoke too high expectations, though (like it would actually take care of computing the standardization). What about just neutrally calling it a `ScaleShiftLayer`, renaming `axes` to `shared_axes`, giving it a `trainable` keyword argument and pointing out your use case in the documentation? Would that work as well or would that be too generic / too difficult to find?

Regarding the implementation, I'm not sure about the `reshape` in the end -- wouldn't it be better to follow the `dimshuffle` pattern as in `BiasLayer`, for example?

**Comment by craffel** [comment]
>  The way it's presented might evoke too high expectations, though (like it would actually take care of computing the standardization).

Yes, that's true; sort of a naming issue.

> What about just neutrally calling it a ScaleShiftLayer,

Or `OffsetScaleLayer` based on my current naming.

> renaming axes to shared_axes,

Sure.

> giving it a trainable keyword argument and pointing out your use case in the documentation?   Would that work as well or would that be too generic / too difficult to find?

If there's a use-case for a generic (potentially trainable) offset-and-scale layer, I'd personally think it would be convenient/helpful to also have a named layer which suggests it is used for standardization and sets trainable to `False` by default.

> Regarding the implementation, I'm not sure about the reshape in the end -- wouldn't it be better to follow the dimshuffle pattern as in BiasLayer, for example?

Hm, I see that.  Is one better than the other in terms of efficiency?  The reshape pattern is more natural to me personally.  The way I have it implemented here works for the use-cases I tested.

**Comment by f0k** [comment]
> Hm, I see that. Is one better than the other in terms of efficiency?

Every symbolic `.shape` introduces a Theano Op, so it might a little less inefficient / more difficult to optimize for Theano (but I don't know).

> The reshape pattern is more natural to me personally.

I think the most natural way would be making the parameters correctly broadcastable right in the constructor and then not care about the shared axes any more. It's just that existing layers already went the route of reducing the dimensionality of their biases (e.g., have a vector for the convolutional layers) and dimshuffling them for applying them, so I followed suit in `BiasLayer`.

> Or `OffsetScaleLayer` based on my current naming.

Fine as well. `ScaleShiftLayer` would actually imply scaling first, so disregard this.

> If there's a use-case for a generic (potentially trainable) offset-and-scale layer

Not sure about actual use cases, it would just feel inconsistent to me to add a layer that is not even trainable by default (because it would be the only one).

> I'd personally think it would be convenient/helpful to also have a named layer which suggests it is used for standardization and sets trainable to `False` by default.

What about a macro function? If we go for a macro function, we could also just add a `ScaleLayer` to `lasagne.layers.special` and add a `trainable` keyword argument to the `BiasLayer` -- or let the macro function remove the `trainable` flag from the layers:
```python
def standardize(layer, offset, scale, axes=..., trainable=False):
    layer = BiasLayer(layer, offset, axes)
    if not trainable:
        layer.params[layer.b].remove('trainable')
    layer = ScaleLayer(layer, scale, axes)
    if not trainable:
        layer.params[layer.scale].remove('trainable')
    return layer
```
Might lead to less code duplication in the long run, and it would be about as convenient to use as your `StandardizationLayer`. What do you think?

**Comment by craffel** [comment]
> I think the most natural way would be making the parameters correctly broadcastable right in the constructor and then not care about the shared axes any more. It's just that existing layers already went the route of reducing the dimensionality of their biases (e.g., have a vector for the convolutional layers) and dimshuffling them for applying them, so I followed suit in BiasLayer.

I don't personally like the idea of the layer changing the shape of the parameters that the user passes to it, but if `BiasLayer` already does it one way and we're sticking to that it's a moot discussion anyways.

> Not sure about actual use cases, it would just feel inconsistent to me to add a layer that is not even trainable by default (because it would be the only one).

Ok.

> What about a macro function? If we go for a macro function, we could also just add a ScaleLayer to lasagne.layers.special and add a trainable keyword argument to the BiasLayer -- or let the macro function remove the trainable flag from the layers:

I am ok with this, with the minor caveat that I don't think there should be a `trainable` argument to the macro - it doesn't make any sense in my mind for the offset/scale to be optimizable when they are used for standardization.

**Comment by f0k** [comment]
> I don't personally like the idea of the layer changing the shape of the parameters that the user passes to it

No no, it doesn't. It was just a question of what dimensionality the parameters are defined as in the first place. My idea was that a `BiasLayer` on 4D input should always have a 4D bias tensor with broadcastable singleton dimensions for all shared axes, instead of creating, say, a 2D tensor and adding the correct broadcastable dimensions in `get_output_for()`. It does the latter now to be compatible with existing biases, but this also means it's got to have the somewhat complex `dimshuffle` pattern generator.

> I am ok with this, with the minor caveat that I don't think there should be a `trainable` argument to the macro - it doesn't make any sense in my mind for the offset/scale to be optimizable when they are used for standardization.

Okay for me as well. One might want to try continuing training from the standardization, but that only makes sense if the next layer doesn't start with a linear transformation anyway. That's probably not going to happen very often.

Do you want to add the `ScaleLayer` and `standardize()` macro (the latter could reuse a lot of your current docstring)? The `ScaleLayer` should probably go into `layers/special.py` after the `BiasLayer`, about `standardize()` I'm not sure.

**Comment by craffel** [comment]
> No no, it doesn't. It was just a question of what dimensionality the parameters are defined as in the first place. My idea was that a BiasLayer on 4D input should always have a 4D bias tensor with broadcastable singleton dimensions for all shared axes, instead of creating, say, a 2D tensor and adding the correct broadcastable dimensions in get_output_for(). It does the latter now to be compatible with existing biases, but this also means it's got to have the somewhat complex dimshuffle pattern generator.

I'll follow the existing convention, if it's something which we decide to change later on I'll leave it to then.

> Do you want to add the ScaleLayer and standardize() macro (the latter could reuse a lot of your current docstring)? The ScaleLayer should probably go into layers/special.py after the BiasLayer, about standardize() I'm not sure.

Sure, I will do this now.

**Comment by craffel** [comment]
Ok, added `ScaleLayer` and `standardize` macro, ready for review.

**Comment by benanne** [comment]
Looks nice :) I had a quick look, didn't do a detailed review. It seems that coveralls is unhappy because the case of `scales=None` is untested.

**Comment by craffel** [comment]
Oops, I actually meant to take that out (I took that case out of `get_output_for`) - I can't really think of a reason anyone would want to use a `ScaleLayer` with no scaling.  If you want the no-scaling, `scales is None` case, I can re-add that back in.

**Comment by benanne** [comment]
No, I think it's fine to get rid of it altogether probably :)

**Comment by craffel** [comment]
Not sure where to put the docs for `standardize`, any ideas are appreciated.

File: lasagne/layers/normalization.py [hunk 1]
```diff
@@ -101,3 +103,58 @@ def get_output_for(self, input, **kwargs):
...
         return input / scale
+
+
+def standardize(layer, offset, scale, shared_axes):
+    """
+    Convenience function for standardizing inputs by applying a fixed offset
+    and scale.  This is usually useful when you want the input to your network
+    to, say, have zero mean and unit standard deviation over the feature
+    dimensions.  This layer allows you to compute those statistics over your
+    training set and include them as part of your network.  The statistics are
```
Comment (f0k) [inline lasagne/layers/normalization.py]: "This layer allows you to compute those statistics" -- sounds as if the layer would do this. Can you rephrase such that it's clear that the layer only allows you to store the normalization offsets/factors within the network, but does not help you with computing them?

File: lasagne/layers/normalization.py [hunk 2]
```diff
@@ -101,3 +103,58 @@ def get_output_for(self, input, **kwargs):
...
+    layer : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    offset : Theano shared variable, expression, numpy array or callable
+        The offset to apply (via subtraction) to the axis/axes being
+        standardized.
+    scale : Theano shared variable, expression, numpy array or callable
+        The scale to apply (via division) to the axis/axes being standardized.
+    shared_axes : 'auto', int or tuple of int
+        The axis or axes which will be shared (i.e. are not being standardized
+        over).
```
Comment (f0k) [inline lasagne/layers/normalization.py]: Mhm, that's not really what it does, is it? `shared_axes` names the axes over which the offset and scale will be shared/broadcasted. It also needs an explanation for `'auto'` (which can be copied over from the `BiasLayer`).

Comment (f0k) [inline lasagne/layers/normalization.py]: Comparing to the `BatchNormLayer`, those axes are actually the axes standardized over, not the axes *not* standardized over. I.e., to compute the offset for `shared_axes=(1,3)`, you would compute `np.mean(X, axis=(1,3))`. It's confusing. I think it may be safer to copy and adapt the explanation from `BiasLayer` or `ScaleLayer`.

File: lasagne/layers/special.py [hunk 3]
```diff
@@ -112,6 +113,67 @@ def get_output_for(self, input, **kwargs):
             return input
 
 
+class ScaleLayer(Layer):
+    """
+    lasagne.layers.ScaleLayer(incoming, scales=lasagne.init.Constant(1),
+    shared_axes='auto', **kwargs)
+
+    A layer that scales its inputs along certain axes.
```
Comment (f0k) [inline lasagne/layers/special.py]: "scale along axes" confuses me -- what about just saying `A layer that scales its inputs by learned coefficients.`? Sharing coefficients over axes is explained fine in the `shared_axes` description.

**Comment by f0k** [comment]
> Not sure where to put the docs for `standardize`, any ideas are appreciated.

If you rebase onto the latest master (sorry about the merge conflict), `layers/normalize.rst` will magically appear for you to put the docs into. I'm still not sure if this function should live in `normalize` or `special`, though (hint: if you move it to `special`, you avoid the merge conflict, if that's a good reason ;) ).

File: lasagne/layers/normalization.py [hunk 4]
```diff
@@ -101,3 +103,58 @@ def get_output_for(self, input, **kwargs):
...
+    Alternatively, to z-score your inputs based on training set statistics, you
+    could set ``offset = training_data.mean(axis=0)`` and
+    ``scale = training_data.std(axis=0)`` instead.
+    """
+    # Subtract the offset
+    layer = BiasLayer(layer, -offset, shared_axes)
+    # Do not optimize the offset parameter
+    layer.params[layer.b].remove('trainable')
+    # Divide by the scale
+    layer = ScaleLayer(layer, 1./scale, shared_axes)
```
Comment (f0k) [inline lasagne/layers/normalization.py]: This is dangerous again -- if `scale` is a numpy array, `1./scale` will upcast it to float64. You could use `T.inv()`, with the fun side effect that it immediately turns a numpy array into a Theano constant that cannot be learned. Alternatively, you could use `lasagne.utils.floatX(1) / scale`.

Aiming for constants might enable some more optimizations for Theano, though (if they can be folded with other operations). We could supply the bias as a constant as well, instead of a shared variable.

File: lasagne/layers/special.py (hunk 3 above)
Comment (craffel) [inline lasagne/layers/special.py]: Ok, changed.

File: lasagne/layers/normalization.py (hunk 1 above)
Comment (craffel) [inline lasagne/layers/normalization.py]: Ok, changed wording.

File: lasagne/layers/normalization.py (hunk 2 above)
Comment (craffel) [inline lasagne/layers/normalization.py]: I understand it in the opposite way (e.g. "standardizing over" means "make these dimensions standardize").  I copied the explanation.

File: lasagne/layers/normalization.py (hunk 4 above)
Comment (craffel) [inline lasagne/layers/normalization.py]: Ok, I switched to using `T.inv`.  I also removed the documentation saying that a callable could be used, which is not true.

**Comment by craffel** [comment]
> If you rebase onto the latest master (sorry about the merge conflict), layers/normalize.rst will magically appear for you to put the docs into. I'm still not sure if this function should live in normalize or special, though (hint: if you move it to special, you avoid the merge conflict, if that's a good reason ;) ).

Ok, I moved it.  Also rebased.  Should be ready for re-review.

File: lasagne/layers/special.py [hunk 5]
```diff
@@ -116,6 +118,124 @@ def get_output_for(self, input, **kwargs):
...
+    achieve this normalization as part of your network, and applies them to its
+    input.  The statistics are supplied as the `offset` and `scale` parameters,
+    which are applied to the input by subtracting `offset` and dividing by
+    `scale`, sharing dimensions as specified by the `shared_axes` argument.
+
+    Parameters
+    ----------
+    layer : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    offset : Theano shared variable, expression, numpy array or callable
```
Comment (f0k) [inline lasagne/layers/special.py]: Cannot be a callable the way it's implemented.

File: lasagne/layers/special.py [hunk 6]
```diff
@@ -116,6 +118,124 @@ def get_output_for(self, input, **kwargs):
...
+        The scale to apply (via division) to the axis/axes being standardized.
+    shared_axes : 'auto', int or tuple of int
+        The axis or axes to share the offset and scale over. If ``'auto'`` (the
+        default), share over all axes except for the second: this will share
+        scales over the minibatch dimension for dense layers, and additionally
+        over all spatial dimensions for convolutional layers.
+
+    Examples
+    --------
+
```
Comment (f0k) [inline lasagne/layers/special.py]: Just for consistency, could you drop this newline, please? Cheers!

File: lasagne/tests/layers/test_special.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 import numpy as np
 import pytest
 import theano
+import lasagne
```
Comment (f0k) [inline lasagne/tests/layers/test_special.py]: I'd prefer `from lasagne.layers import InputLayer, standardize, get_output` down in `test_standardize()` instead of a global import. (But also just for consistency, because most test suites were written to avoid a global `lasagne` import -- I don't know who started it. A possible advantage is that it's easier to track which Lasagne components specific test suites depend on.)

File: lasagne/tests/layers/test_special.py [hunk 8]
```diff
@@ -219,6 +220,79 @@ def test_undefined_shape(self, BiasLayer):
...
+    l_std = lasagne.layers.standardize(
+        l_in, X.min(axis=0), (X.max(axis=0) - X.min(axis=0)), shared_axes=0)
+    out = lasagne.layers.get_output(l_std).eval({l_in.input_var: X})
+    assert np.allclose(out.max(axis=0), 1.)
+    assert np.allclose(out.min(axis=0), 0.)
+    # More complicated example
+    X = np.random.standard_normal(
+        (50, 3, 100, 10)).astype(theano.config.floatX)
+    mean = X.transpose(0, 2, 1, 3).mean(axis=0).mean(axis=0)
+    std = X.transpose(0, 2, 1, 3).std(axis=0).std(axis=0)
```
Comment (f0k) [inline lasagne/tests/layers/test_special.py]: It would be easier to follow as:
```python
mean = X.mean(axis=(0,2))
std = X.std(axis=(0, 2))
```
Unless you had something specific in mind when computing the standard deviation of standard deviations?

**Comment by f0k** [comment]
> Ok, I moved it. Also rebased. Should be ready for re-review.

Thanks, looks very good except for a few comments!

File: lasagne/layers/special.py (hunk 5 above)
Comment (craffel) [inline lasagne/layers/special.py]: Oops, missed that one, fixed.

File: lasagne/layers/special.py (hunk 6 above)
Comment (craffel) [inline lasagne/layers/special.py]: Ok.

File: lasagne/tests/layers/test_special.py (hunk 7 above)
Comment (craffel) [inline lasagne/tests/layers/test_special.py]: Ok.

File: lasagne/tests/layers/test_special.py (hunk 8 above)
Comment (craffel) [inline lasagne/tests/layers/test_special.py]: Yes, this is better, changed.

**Comment by craffel** [comment]
Ok, addressed your comments.

**Comment by f0k** [comment]
Nice! Will merge tomorrow if there are no further comments.

**Comment by craffel** [comment]
Bump!

**Comment by benanne** [comment]
I'm not @f0k, but I'll merge it as there are no further comments :) Thanks!
